### ISSUE TITLE
Passing undefined to node.log throws TypeError

### ISSUE THREAD

**Issue by M0ebiu5** [node-red/node-red#1418]
I'm getting a strange error "TypeError: Cannot read property 'stack' of undefined" after the flow is running some time - or sometimes after changing and deploying another node.

I've found the problem cause the node "prep values" did not output anything after some time - but it also didnt show any error. To trace the problem i've inserted the line "node.warn(last.payload.samplecnt);". Now, node-red throws the above error.

Here is a sample flow, to reproduce the error - it occurs after a certain time (sometimes it starts to fail also after a change (eg: blank only) in "check heating")

Environment: linux, node version 8.2.1, nr version 0.17.5

`[{"id":"1ee2de46.19af92","type":"join","z":"96082429.7476f8","name":"","mode":"custom","build":"array","property":"payload","propertyType":"msg","key":"topic","joiner":"\\n","joinerType":"str","accumulate":false,"timeout":"","count":"3","x":408,"y":1606,"wires":[["9cff06fc.33ff78"]]},{"id":"7ec1a885.306188","type":"inject","z":"96082429.7476f8","name":"","topic":"x","payload":"25.7","payloadType":"str","repeat":"1","crontab":"","once":false,"x":208,"y":1547,"wires":[["1ee2de46.19af92"]]},{"id":"89ca2b25.034fd8","type":"inject","z":"96082429.7476f8","name":"","topic":"y","payload":"23.7","payloadType":"str","repeat":"1","crontab":"","once":false,"x":193,"y":1591,"wires":[["1ee2de46.19af92"]]},{"id":"8b8bee8e.eafc2","type":"inject","z":"96082429.7476f8","name":"","topic":"z","payload":"24.4","payloadType":"str","repeat":"4","crontab":"","once":false,"x":192,"y":1656,"wires":[["1ee2de46.19af92"]]},{"id":"9cff06fc.33ff78","type":"function","z":"96082429.7476f8","name":"prep values","func":"var last = context.get(\"lastx\");\nvar tst = new Date().getTime();\n\nvar vor = parseFloat(msg.payload[1]);\nvar rue = parseFloat(msg.payload[0]);\nvar solar = parseFloat(msg.payload[2]);\n\nmsg.payload = {};\n//node.warn(\"1\");\nmsg.payload.vor = vor;\n//node.warn(\"2\");\nmsg.payload.rue = rue;\nmsg.payload.vordiff = 0;\nmsg.payload.ruediff = 0;\nmsg.payload.tst = tst;\n\nmsg.payload.solar = solar;\nmsg.payload.solardiff = 0;\n\nif (typeof last !== \"undefined\") {\nnode.warn(last.payload.samplecnt);    \n    msg.payload.samplecnt = last.payload.samplecnt;\n    if (last.payload.samplecnt > 10) {\n        msg.payload.vordiff = vor - last.payload.vor;\n        msg.payload.ruediff = rue - last.payload.rue;\n        msg.payload.solardiff = solar - last.payload.solar;\n        msg.payload.samplecnt = 0;\n        context.set(\"lastx\",msg);\n    } else {\n        last.payload.samplecnt++;\n        context.set(\"lastx\",last);\n        return null;\n    }\n} else {\n    msg.payload.samplecnt = 0;\n    context.set(\"lastx\",msg);\n}\n\nreturn msg;","outputs":1,"noerr":0,"x":574,"y":1653,"wires":[["289695ff.3ffc0a","8168f5b2.a55b58"]]},{"id":"289695ff.3ffc0a","type":"function","z":"96082429.7476f8","name":"check heating","func":"var hotwaterdelay = 63; //\n\nvar heat = context.get(\"heating\") || {\"wait\": new Date().getTime(),\"state\":\"undef\",\"reason\":\"default\",\"tst\": new Date()};\nvar oldstate = heat.state;\n\n//prep status msg\nif (typeof msg.payload.status !== \"undefined\") {\n    msg.payload.status = {};\n    for (var att in heat) {\n            if (heat.hasOwnProperty(att)) { \n                msg.payload.status[att] = heat[att];\n                if (att == \"tst\")\n                    msg.payload.status[att] = new Date();\n            }\n    }\n    return msg;\n}\n\nif (heat.wait > new Date().getTime())\n    return null;\n    \nnode.warn(msg.payload);\n\nif (msg.payload.solar > 28 && msg.payload.rue < 25) {\n    heat.state = \"on\";\n    heat.reason = \"solar power available\";\n}\n\nif (msg.payload.vordiff > 0.7 || msg.payload.vor > 30) {\n    heat.state = \"off\";\n    heat.wait = new Date().getTime() + (hotwaterdelay * 60000);   \n    heat.reason = \"hot water on\";\n}\n\nif (msg.payload.rue > 25) {\n    heat.state = \"off\";\n    heat.reason = \"floor temp > 25\";\n}\ncontext.set(\"heating\",heat);\n\nif (heat.state === oldstate)\n    return null;\n\nmsg.topic = \"so/cmnd/Heating/Power\"\nmsg.payload = heat.state;\nmsg.reason = heat.reason;\nreturn msg;","outputs":1,"noerr":0,"x":761,"y":1706,"wires":[[]]},{"id":"d52c4968.925018","type":"inject","z":"96082429.7476f8","name":"","topic":"status","payload":"{\"status\":\"\"}","payloadType":"json","repeat":"","crontab":"","once":false,"x":540,"y":1725,"wires":[["289695ff.3ffc0a"]]},{"id":"8168f5b2.a55b58","type":"debug","z":"96082429.7476f8","name":"","active":true,"console":"false","complete":"true","x":1017,"y":1325,"wires":[]}]`

**Comment by knolleary** [comment]
Please provide the full error including stack if any.

**Comment by knolleary** [comment]
Put a try/catch around the code in the Function node in question ...
```
28 Sep 17:45:20 - [error] [function:prep values] TypeError: Cannot read property 'stack' of undefined
        at consoleLogger [as handler] (/Users/nol/code/node-red/node-red/red/runtime/log.js:77:31)
        at null.<anonymous> (/Users/nol/code/node-red/node-red/red/runtime/log.js:61:18)
        at emitOne (events.js:77:13)
        at emit (events.js:169:7)
        at /Users/nol/code/node-red/node-red/red/runtime/log.js:133:21
        at Array.forEach (native)
        at Object.module.exports.log (/Users/nol/code/node-red/node-red/red/runtime/log.js:132:21)
        at log_helper (/Users/nol/code/node-red/node-red/red/runtime/nodes/Node.js:231:9)
        at FunctionNode.Node.warn (/Users/nol/code/node-red/node-red/red/runtime/nodes/Node.js:239:5)
        at Object.FunctionNode.sandbox.__node__.warn (/Users/nol/code/node-red/node-red/nodes/core/core/80-function.js:92:31)
        at evalmachine.<anonymous>:23:6
        at evalmachine.<anonymous>:46:3
        at FunctionNode.<anonymous> (/Users/nol/code/node-red/node-red/nodes/core/core/80-function.js:197:33)
        at emitOne (events.js:77:13)
        at FunctionNode.emit (events.js:169:7)
        at FunctionNode.Node.receive (/Users/nol/code/node-red/node-red/red/runtime/nodes/Node.js:215:14)
        at JoinNode.Node.send (/Users/nol/code/node-red/node-red/red/runtime/nodes/Node.js:137:22)
        at JoinNode.completeSend (/Users/nol/code/node-red/node-red/nodes/core/logic/17-split.js:311:18)
        at JoinNode.<anonymous> (/Users/nol/code/node-red/node-red/nodes/core/logic/17-split.js:461:21)
        at emitOne (events.js:77:13)
        at JoinNode.emit (events.js:169:7)
        at JoinNode.Node.receive (/Users/nol/code/node-red/node-red/red/runtime/nodes/Node.js:215:14)
        at InjectNode.Node.send (/Users/nol/code/node-red/node-red/red/runtime/nodes/Node.js:137:22)
        at InjectNode.<anonymous> (/Users/nol/code/node-red/node-red/nodes/core/core/20-inject.js:64:22)
        at emitOne (events.js:77:13)
        at InjectNode.emit (events.js:169:7)
        at null._repeat (/Users/nol/code/node-red/node-red/nodes/core/core/20-inject.js:37:22)
        at wrapper [as _onTimeout] (timers.js:279:11)
        at Timer.listOnTimeout (timers.js:92:15)
```

**Comment by knolleary** [comment]
Ahh - this is caused by passing `undefined` to any of the `node.log`/`warn`/`error` functions.

We need to fix that. In the meantime, avoid passing `undefined` to those functions ;)

**Comment by M0ebiu5** [comment]
Ok, but i dont see how msg.payload.samplecnt could be undefined - can you give me a hint?

**Comment by knolleary** [comment]
@M0ebiu5 the line that is failing is:

```
node.warn(last.payload.samplecnt); 
```
Tracing that back, `last` is set by `context.get("lastx")`

Sticking a `console.log` to see what `last` actually is shows its:

```
{ topic: 'so/cmnd/Heating/Power',
      payload: 'off',
      _msgid: '30914c1f.cf6eb4',
      reason: 'floor temp > 25' }
```

So, `last.payload` is a String - so you cannot set properties on it which is why `msg.payload.samplecnt` is undefined.

As you're storing the whole msg, there's no need to add your extra properties under `msg.payload` - just set them on `msg` directly.

**Comment by M0ebiu5** [comment]
sorry, but i dont get it....
The line 
`node.warn(last.payload.samplecnt);` 
is from the node "prep values".

But the output
`{ topic: 'so/cmnd/Heating/Power',
    payload: 'off',
     _msgid: '30914c1f.cf6eb4',
     reason: 'floor temp > 25' }` 
is a msg from the node "check heating". How can this value be "last" in the node "prep values"?

From my point of view, the error occurs in "prep values". There, i prepare a msg.payload object with the attribute "samplecnt". At some point i want to reset the samplecount with:
`        msg.payload.samplecnt = 0;
        context.set("lastx",msg);`

But after reentering the node and restoring the lastx object, the samplecnt attribute is suddenly undefined. (i just traced the type before storing into context and retrieving).

**Comment by knolleary** [comment]
You are storing a reference to a msg object in context. You then modify that msg in a subsequent node which modifies what is stored in context.

If you end `prep values` function with:
```
return RED.util.cloneMessage(msg);
```
it will break the reference between the msg passed on and what you've stored in context.

**Comment by M0ebiu5** [comment]
Oh, i thought that each nodes context only belongs to himself and is not accessible from "outside" (unlike flowContext). 
Thanks!

**Comment by knolleary** [comment]
well, technically they are separately... but if you store a reference to a message, and then pass that message on, it can leak.

**Comment by knolleary** [comment]
going to reopen this as we need to address the error of passing `undefined` to the logging functions.

**Comment by knolleary** [comment]
Fixed in 0.18 branch
